File: 0009-palindrome-number/0009-palindrome-number.cpp
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0) return false;
        return x == rev(x);
    }

private:
    long long rev(int x) {
        if (x == 0) return 0; // Handle the special case for 0
        int digits = static_cast<int>(log10(x)) + 1;
        return helper(x, digits);
    }

    long long helper(int n, int digits) {
        if (n % 10 == n) {
            return n;
        }
        int rem = n % 10;
        return rem * static_cast<long long>(pow(10, digits - 1)) + helper(n / 10, digits - 1);
    }
};

File: 0009-palindrome-number/0009-palindrome-number.py
import math

class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False
        return x == self.rev(x)

    def rev(self, x: int) -> int:
        if x == 0:  # Handle the special case for 0
            return 0
        digits = int(math.log10(x)) + 1
        return self.helper(x, digits)

    def helper(self, n: int, digits: int) -> int:
        if n % 10 == n:
            return n
        rem = n % 10
        return rem * (10 ** (digits - 1)) + self.helper(n // 10, digits - 1)
